src/Microsoft.AspNetCore.Hosting/Startup/StartupBase.cs

```diff
@@ -16,4 +16,25 @@ public virtual IServiceProvider ConfigureServices(IServiceCollection services)
             return services.BuildServiceProvider();
         }
     }
+
+    public abstract class StartupBase<TContainer> : IStartup
+    {
+        private readonly IServiceProviderFactory _factory;
+
+        public StartupBase(IServiceProviderFactory factory)
+        {
+            _factory = factory;
+        }
+
+        public abstract void Configure(IApplicationBuilder app);
+
+        public IServiceProvider ConfigureServices(IServiceCollection services)
+        {
+            var builder = _factory.CreateBuilder(services);
```

Comment:
Have you considered what I mentioned in https://github.com/aspnet/Hosting/pull/833#discussion_r74199132?